server/data_common/matrix_loader.py

```diff
@@ -0,0 +1,173 @@
+
+    def __init__(self):
+        # key is location, value is tuple of (MatrixDataCacheItem, last_accessed)
+        self.datasets = {}
+        self.lock = threading.Lock()
+
+    @contextmanager
+    def data_adaptor(self, location, app_config):
+        # create a loader for to this location if it does not already exist
+        with self.lock:
+            value = self.datasets.get(location)
+            if value is not None:
+                cache_item = value[0]
+                last_accessed = time.time()
+                self.datasets[location] = (cache_item, last_accessed)
+            else:
+                while True:
+                    # find the last access times for each loader
+                    items = list(self.datasets.items())
+                    sorted(items, key=lambda x: x[1][1])
+                    if len(items) < self.MAX_CACHED:
+                        break
+
+                    # close the least recently used loader
+                    oldest = items[0]
+                    oldest_cache = oldest[1][0]
+                    oldest_key = oldest[0]
+                    oldest_cache.delete()
+                    del self.datasets[oldest_key]
+
+                last_accessed = time.time()
+                loader = MatrixDataLoader(location)
+                cache_item = MatrixDataCacheItem(loader)
+                self.datasets[location] = (cache_item, last_accessed)
+        try:
+            data_adaptor = cache_item.acquire(app_config)
+            yield data_adaptor
+        finally:
+            cache_item.release()
+
+
+class MatrixDataType(Enum):
+    H5AD = "h5ad"
+    TILEDB = "tiledb"
+    UNKNOWN = "unknown"
+
+
+class MatrixDataLoader(object):
+
+    def __init__(self, location, etype=None):
+        self.location = location
+        if etype is None:
+            self.etype = self.matrix_data_type()
+        else:
+            self.etype = etype
+        self.matrix_type = None
+        if self.etype == MatrixDataType.H5AD:
+            from server.data_scanpy.scanpy_adaptor import ScanpyAdaptor
+            self.matrix_type = ScanpyAdaptor
+        elif self.etype == MatrixDataType.TILEDB:
+            from server.data_tiledb.tiledb_adaptor import TileDbAdaptor
+            self.matrix_type = TileDbAdaptor
+
+    def matrix_data_type(self):
+        if self.location.endswith(".h5ad"):
+            return MatrixDataType.H5AD
+        elif ".cxg" in self.location:
+            return MatrixDataType.TILEDB
+        else:
+            return MatrixDataType.UNKNOWN
+
+    def pre_load_validation(self):
+        if self.etype == MatrixDataType.UNKNOWN:
+            raise RuntimeError(f"{self.location} does not have a recognized type: .h5ad or .cxg")
+        self.matrix_type.pre_load_validation(self.location)
+
+    def file_size(self):
+        return self.matrix_type.file_size(self.location)
+
+    def open(self, app_config):
+        try:
+            # create and return an object to the matrix_type
```

Comment:
nit: comment is misleading.   It returns the DataAdaptor

Reply:
fixed